[1 - 3단계 방탈출 사용자 예약] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
https://prolog.techcourse.co.kr/studylogs
## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.
- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->
안녕하세요 이프~
장기 미션에 이어 또 뵙게 되어 기쁩니다ㅎㅎ

지난 번 이프가 타인으로부터 얻을 수 있는 학습 기회를 놓치지 않았으면 좋겠다고 말씀해주셨었는데요.
그 뒤로 다른 사람들의 코드와 의견을 많이 접해보면서 학습 기회를 많이 만들어보려고 노력중입니다.
같은 이유로 이번 페어 프로그래밍에서도 페어의 코드에서 시작했습니다.
좋은 조언 감사드립니다.😄

그럼 이번 리뷰도 잘부탁드립니다~!


[커밋 링크](https://github.com/woowacourse/spring-roomescape-member/pull/199/files/2b94bbe1df72e8a56f45fab4bbe29f5b55d63858..66550f1f6753b47afdbf91c4095af20245b44f7c)입니다!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

solarsalt committed:
docs: add PR template

yesjuhee committed:
chore: 미션1 코드 마이그레이션

yesjuhee committed:
refactor: API 응답 상태 코드 변경

yesjuhee committed:
docs: 1단계 요구사항 정리

yesjuhee committed:
feat: 예약자명은 최소 1글자, 최대 5글자여야한다

yesjuhee committed:
feat: `HH:mm` 형식으로 시간을 입력해야한다

yesjuhee committed:
feat: 예약 생성 요청 형식 검증한다

yesjuhee committed:
feat: 같은 날짜와 시간에 중복 예약을 할 수 없다

yesjuhee committed:
refactor: Reservation 객체 생성 방식 변경

kjyyjk committed:
feat: 예약 시간을 중복 생성할 수 없다

kjyyjk committed:
feat: 존재하는 시간 ID를 입력해야한다

yesjuhee committed:
feat: 특정 시간에 대한 예약이 존재할 때 예약 시간을 삭제할 수 없다

yesjuhee committed:
feat: 과거에 대한 예약은 불가능하다

kjyyjk committed:
refactor: controller 리팩터링

- 전역적으로 사용되는 ExceptionHandler를 ControllerAdvice로 분리했다.

kjyyjk committed:
refactor: ReservationDateTime 추가

kjyyjk committed:
refactor: Dao 인터페이스 분리

yesjuhee committed:
test: 테스트용 페이크 객체 생성

yesjuhee committed:
test: ReservationService 테스트 작성

kjyyjk committed:
test: ReservationTimeService 테스트 작성

kjyyjk committed:
refactor: 예약자명을 MemberName으로 분리

yesjuhee committed:
docs: 2단계 요구사항 정리

yesjuhee committed:
feat: 테마 관리 페이지 응답 API 추가

yesjuhee committed:
feat: 테마 DDL 추가

kjyyjk committed:
feat: Theme 도메인 객체 추가

yesjuhee committed:
feat: 테마 생성 API 구현

kjyyjk committed:
feat: 테마 생성 API 컨트롤러 구현

kjyyjk committed:
feat: 테마 목록 조회 API 구현

yesjuhee committed:
refactor: 예약 API에 테마 적용

kjyyjk committed:
feat: 테마 삭제 API 구현

yesjuhee committed:
refactor: 컨트롤러 응답 헤더 리팩터링

yesjuhee committed:
chore: 응답 HTML 변경, 경고창 추가

kjyyjk committed:
feat: 사용자 화면 응답 API 추가

yesjuhee committed:
docs: 예약 가능 시간 API 요구사항 추가

yesjuhee committed:
feat: 예약 가능 시간 조회 API 추가

kjyyjk committed:
docs: 인기 테마 조회 API 요구사항 추가

kjyyjk committed:
chore: 목업 데이터 추가

kjyyjk committed:
feat: 인기 테마 조회 API 구현

yesjuhee committed:
feat: 검증 로직 추가

kjyyjk committed:
refactor: Service, Dao 리팩터링

kjyyjk commented on src/main/java/roomescape/controller/ThemeController.java:
```diff
@@ -0,0 +1,48 @@
+        themeService.deleteThemeById(id);
+        return ResponseEntity.noContent().build();
+    }
+
+    @GetMapping("/popular-themes")
+    public ResponseEntity<List<ThemeResponse>> findPopularThemes() {
+        List<ThemeResponse> responses = themeService.findPopularThemes();
+        return ResponseEntity.ok().body(responses);
+    }
```
페어와 함께 API를 설계할 때 고민을 많이 했었던 부분입니다.
처음에는 /themes/popular가 자연스럽다고 생각했습니다.
하지만 REST API 설계 시 명사를 사용을 지향한다는 점과 케밥 케이스를 지향한다는 것을 알게 되어 
`/popular-themes`라고 지었습니다!

sinb57 replied:
REST API는 무엇일까요? `/themes/popular`가 자연스러움에도 REST API의 규칙을 따를 필요가 있는지 궁금해요~

kjyyjk replied:
`RESTful API`는 REST 지침에 따라 사용자와 자원의 상태를 주고 받는 인터페이스입니다.

제가 생각하는 좋은 API는 일관적이며 의도가 명확히 드러나야 합니다.
REST와 API 모두 어떤 규약이 아니기 때문에 꼭 지켜야 하는 것은 아닙니다.
하지만 컨벤션을 따름으로써 일관성을 가지며 그 의도를 파악하기가 쉬워진다고 생각합니다.

저는 현재 어떤 팀에 속해 있지 않아 별도의 컨벤션이 존재하지 않으니, 일반적으로 많이 쓰이는 컨벤션을 따르면 좋을 것 같아요.

그렇기에 명사를 사용해 계층을 표현할 필요가 있다고 생각했습니다.
다시 한번 URI를 고민해봤고, `/themes/popular-themes`가 적절할 것 같다고 생각했습니다.

반대로 팀 컨벤션이 존재한다면 그에 따를 것 같습니다.

sinb57 replied:
기준이 명확해서 좋네요ㅎ API 명세는 개개인의 취향에 따라 다르기 때문에 정답은 없어요~ 말씀하신대로 팀 컨벤션을 따르는 것이 옳고, 경우에 따라선 팀 컨벤션을 새로 제안할 수도 있지요

 저는`/themes/popular`라고 지었을 것 같아요. 그러면 controller`에 `@RequestMapping("/themes")`를 두어 path 중복을 제거할 수 있으니까요~

kjyyjk commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,66 @@
+        final ReservationTime reservationTime = findReservationTime(reservationRequest.timeId());
+        final Theme theme = findTheme(reservationRequest.themeId());
+        final Reservation reservation = reservationRequest.convertToReservation(reservationTime, theme);
+        if (reservation.isBefore(LocalDateTime.now())) {
+            throw new IllegalArgumentException("지나간 날짜와 시간은 예약 불가합니다.");
+        }
```
service에서 reservation 객체의 예약 시간이 과거 시간인지 확인하고, 참이라면 예외를 던지고 있습니다.
이 부분에서 해당 예외를 던지는 것 또한 Reservation 내부에서 하면 좋지 않을까?하고 생각했습니다.

두 방식 모두 적절한 것 같지만 아직까지는 무엇이 비즈니스 로직/도메인 로직이고 어디에 위치하는게 적절한 것인가에
관한 기준이 명확하지 않아 생각만 해보고 있습니다.

현재까지는 다음과 같은 의견을 가지고 있습니다.
같은 시간에 중복되는 예약이 있는 지를 검증하는 것은 `db에 의존`해야하기 때문에 `service`에서 처리하고,

예약 시간이 과거 시간인지를 검증하는 것 자체는 Reservation 내부에서 처리할 수 있기 때문에 `도메인 객체`에서 검증하는 것이 적절하다고 생각합니다.(현재 시간 정보는 메서드 파라미터로 주입)

sinb57 replied:
> 현재까지는 다음과 같은 의견을 가지고 있습니다.

현재 책임 분리를 잘 하셨다고 생각해요. 도메인으로 책임을 적절히 분배했어요.👍

> 아직까지는 무엇이 비즈니스 로직/도메인 로직이고 어디에 위치하는게 적절한 것인가에 관한 기준이 명확하지 않아 생각만 해보고 있습니다.

레오는 `Service`의 역할과 책임이 무엇이라고 생각하는지 궁금해요~

kjyyjk replied:
막연하게 생각했을 때 서비스가 가지는 책임은 비즈니스 로직의 흐름 관리였습니다.

조금 더 깊게 생각해보기 위해 `비즈니스 로직이 무엇인가`부터 물음을 던지기 시작했고 다음과 같은 결론을 내렸습니다.
```
비즈니스 로직은 사용자의 요구를 해결하기 위한 핵심 로직이다.(요청과 요구는 다르다. 요청이 요구를 포함하고 있다.)
비즈니스 로직은 도메인 로직과 도메인 로직이 아닌 것으로 구분할 수 있다.

도메인 로직은 도메인 객체들의 관심사이자 스스로 해결할 수 있는 로직이다.
따라서 도메인 내부에 위치한다.

도메인 로직이 아닌 것은 외부에 의존성을 가지는 로직이다.
이는 서비스 계층에 위치한다.

예를 들어 dao나 controller는 외부와 통신하는 책임을 가진다.
이에 의존하는 것은 곧 외부에 대한 의존성을 가지는 것이다.
도메인 객체는 외부에 관심이 없기 때문에 이에 대한 의존성을 가질 필요가 없다.
대신 Service가 가지게 된다.
```
따라서 제가 생각하는 서비스의 책임은 `외부 의존 로직 수행 및 비즈니스 로직의 흐름 관리`입니다.
이렇게만 보면 또 `외부 의존 로직 수행`과 `전체적인 흐름 관리` 책임을 각각 다른 객체로 나눌 수도 있을 것 같아요.

sinb57 replied:
처음 질문에 대한 답변을 잘못했어서, DM으로 답변드린 내용을 히스토리 차원에서 남겨보아요~
<img width="900" alt="image" src="https://github.com/user-attachments/assets/3f8db367-1932-47c0-8edb-c539d1ad35f7" />

sinb57 replied:
Service에 대한 설명 감사해요~ 말씀하신대로 Service 계층은 외부 의존을 처리하기 위함이 주를 이뤄요. 외부 의존이 도메인까지 내려가지 않기 때문에, 도메인은 외부 의존 변경에 자유롭고 더욱 견고한 로직을 만들어 관리할 수 있게 돼요.

대체로 모든 아키텍쳐와 설계, 코드는 변경의 관심사에 중점을 두고 바라보면 이해하기 쉬울 거예요.

kjyyjk commented on src/main/java/roomescape/service/reservation/Theme.java:
```diff
@@ -0,0 +1,43 @@
+        return name.getName();
+    }
+
+    public ThemeName getThemeName() {
+        return name;
+    }
```
ThemeName은 테마 이름을 관리하는 책임을 지니고 있습니다.
그리고 외부에서 주입 받는 것이 아닌, Theme 생성자 내부에서 생성되어 Theme와 같은 생명 주기를 가지고 있습니다.

그렇기 때문에 ThemeName은 Theme 내부에 캡슐화되어야할 정보이고,
외부에서는 Theme가 ThemeName을 가지고 있는 지를 가급적 몰라야한다는 생각인데요!

하지만 `ThemeDao#isExists`에 ThemeName을 파라미터로 넘기는 것이 적절해보인다는
페어의 의견에 따라 현재처럼 `getThemeName`을 만들었습니다.

만약 제가 수정을 한다면 외부에서는 `getName`을 통해 string만 받아볼 수 있게끔 할 것 같습니다. 

혹시 제가 잘못된 방향으로 캡슐화에 매몰된 것인지..? 궁금합니다.

sinb57 replied:
좋은 고민이에요! 값객체를 사용하다보면 자연스레 찾아오는 순간이네요ㅎ

왜 외부에서는 Theme가 ThemeName을 가지고 있는지 몰라야 한다고 보시나요? getter는 주로 언제 사용되어야 할까요?

kjyyjk replied:
ThemeName은 Theme의 이름을 검증하고 관리하는 책임을 가진 값 객체입니다.
현재 방탈출 예약 관련 시스템 상 이런 책임을 필요로 하는 곳은 Theme 뿐입니다.
그렇기에 외부에 ThemeName을 반환하는 것은 단순히 이름 전달을 위한, 데이터 중심 사고라고 생각했습니다.
이는 외부에 불필요한 의존성을 만들고 결합도를 증가시킵니다.

따라서 ThemeName을 가지고 있는 것은 Theme 내부에 캡슐화되어야 할 내부 구현이며,
이를 노출할 경우 도메인 내부 구현의 변경이 service, dao까지 전파될 것이라고 판단했습니다.

kjyyjk replied:
getter는 주로 뷰나 db에게 데이터를 있는 그대로 전달해야 하는 경우에 사용되어야합니다.
그 외의 경우에는 도메인 객체가 스스로 처리할 수 있게끔 메세지를 던져야한다고 생각합니다.

sinb57 replied:
ThemeName의 생애주기는 Theme 안에서 시작하고 안에서 끝나야 한다는 의견에 동의하지 않아요. 때론 ThemeName이 Theme와 독립적으로 존재할 수도 있어요.

상황 하나를 예로 들어볼게요. Theme 생성 시 랜덤으로 Theme 이름을 지정하는 기능을 만들어야 해요. 이 때 Theme 이름에 대한 규칙은 ThemeName가 검증해주기 때문에, Theme가 아닌 ThemeName을 활용할 수 있어요.

값객체가 내부에서만 존재해야 한다는 의견에 극구 반대하는 것은 아니지만, 꼭 그래야 하는 것은 아니라고 말씀드리고 싶어요. 내부에서만 존재하는 것이 옳을 때가 많겠지만, 그렇지 않은 경우에는 유연하게 접근할 필요가 있다는 것을 말씀드리고 싶네요.

kjyyjk replied:
말씀주신 것처럼 도메인이 가지는 값 객체를 어떻게 다룰 것인가는 상황에 맞춰 유연하게 가져갈 수 있으면 좋겠네요!
좋은 말씀 감사드립니다.

다만 현재로써는 예시 같은 상황이 존재하지 않는 거 같아요.🤔
그렇기 때문에 우선 내부에서만 사용하게 하고 필요한 때가 온다면 밖에서 사용하는 방식은 어떤지 궁금합니다!
이는 변경에 유연하지 못한 것일까요?

지난 장기 미션에서 미래를 예측하여 미리 가져간 구조는 변경에 유연한 것과 다르다고 피드백 주셨던 것이
기억에 남아 여쭤봅니다.

sinb57 replied:
맞아요~ 지금은 예시와 같은 상황이 없죠. 그렇다고 해당 상황을 가정하여 변경해야 할 코드도 없어요. 때문에 미래를 예측한 오버 엔지니어링은 아니라고 생각해요.

코멘트를 남겼던 이유와 위와 같은 예시를 남겼던 이유는, 레오가 `ThemeName은 Theme 내부에 캡슐화되어야할 정보이고, 외부에서는 Theme가 ThemeName을 가지고 있는 지를 가급적 몰라야한다는 생각인데요!`라고 말슴주셨기 때문이에요. 값객체를 무조건 숨겨야 한다고 생각하신 듯 하여, 꼭 그렇지 않은 않다고 말씀드리기 위해 시작된 코멘트였답니다~

kjyyjk commented on src/main/java/roomescape/repository/H2ReservationDao.java:
```diff
@@ -0,0 +1,152 @@
+    public static final String RESERVATION_TABLE = "reservation";
+    public static final String RESERVATION_PK = "id";
+
+    private final JdbcTemplate jdbcTemplate;
```
?와 인덱스 기반 데이터 전달보다는 파라미터명으로 전달하는 것이 더 안전하며 읽기 좋다고 생각하기 때문에
1차 피드백 이후 `NamedParameterJdbcTemplate`으로 수정할 예정입니다!

kjyyjk replied:
그랬었지만 이미 이전 미션에서 사용해봤기 때문에 다른 부분에서 피드백을 우선 빠르게 주고 받는게 좋다고 판단해서 추후 변경해보겠습니다..!

kjyyjk replied:
수정했습니다!

kjyyjk reviewed (commented):
리뷰 잘부탁드립니다!

kjyyjk committed:
refactor: 사용하지 않는 코드 삭제

sinb57 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -23,19 +26,20 @@ public ReservationController(final ReservationService reservationService) {
-            @RequestBody final ReservationRequest reservationRequest
-    ) {
-        return reservationService.createReservation(reservationRequest);
+    public ResponseEntity<ReservationResponse> create(@RequestBody @Valid final ReservationRequest reservationRequest) {
```
`@Valid`는 어떤 역할을 하나요?

kjyyjk replied:
`@Valid`가 붙은 객체의 내부에 존재하는 Constraint 애너테이션 기반으로 객체를 검증합니다.
만약 검증 조건에 걸린다면 `MethodArgumentNotValidException`을 발생합니다.
검증 객체의 다음 인자로 BindingResult를 받으면 검증 시 발생한 에러를 잡아두고 메서드 내부에서 다룰 수 있습니다.

상황에 따른 조건부 검증이 필요하다면 `@Valid` 대신 `@Validated`의 그룹 속성을 활용할 수 있습니다.

sinb57 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -23,19 +26,20 @@ public ReservationController(final ReservationService reservationService) {
-        return reservationService.createReservation(reservationRequest);
+    public ResponseEntity<ReservationResponse> create(@RequestBody @Valid final ReservationRequest reservationRequest) {
+        final ReservationResponse response = reservationService.createReservation(reservationRequest);
+        return ResponseEntity.created(URI.create("/reservations/" + response.id())).body(response);
```
`create` 상태코드는 왜 URI가 필요할까요??

kjyyjk replied:
`@ResponseStatus(HttpStatus.CREATED)`로 하면 URI가 필요 없고, `ResponseEntity`를 사용하면 URI가 필요합니다.
이로부터 `created` 응답에 URI가 필수 요소는 아니라는 것을 알게 됐습니다.

그러면 ResponseEntity는 왜 URI를 요구하는지 생각해보았고, 자주 쓰이기 때문에 편의성을 위함이라고 생각했습니다.
 

created는 요청에 따른 리소스가 생성되었을 때 반환하는 응답 코드이고, Location은 redirect 경로를 지정하는 Http Response의 header입니다. 그리고 대부분의 웹 애플리케이션에서는 리소스 생성 이후 해당 리소스가 위치한 곳으로 이동합니다.
따라서 ResponseEntity는 편의성을 위해 `created`를 사용함과 동시에 Location Header에 담을 URI를 필요로 한다고 생각합니다.

sinb57 commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,71 @@
+        String path = request.getServletPath();
+        if (request.getQueryString() != null) {
+            path += "?" + request.getQueryString();
+        }
```
메서드 분리를 통해 중복을 제거해보면 어떨까요?

sinb57 commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,71 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
의도치 않게 발생하는, 미처 예외처리하지 못한 에러에 대해서는 대응이 필요 없을까요?

kjyyjk replied:
미처 잡지 못한 예외는 Exception으로 잡아 처리하게끔 수정했습니다!
fcfd3b28df13d7db41e4a841d31959eb291fecc7

sinb57 commented on src/main/java/roomescape/repository/H2ReservationDao.java:
```diff
@@ -0,0 +1,152 @@
+import roomescape.service.reservation.ReservationTime;
+import roomescape.service.reservation.Theme;
+
+@Repository
+public class H2ReservationDao implements ReservationDao {
```
`Repository`와 `Dao`는 무슨 차이가 있나요? 각각 어떤 정의를 지녔는지 궁금해요~

kjyyjk replied:
Dao는 데이터베이스의 관점에서 바라본 것으로 테이블과 일대일 매핑됩니다.
반면 Repository는 객체 관점에서 바라본 것으로 객체 저장소에 가깝습니다.

오며가며 크루들에게서 듣고 이번 피드백에 답하기 위해 직접 찾아본 내용인데,
아직 와닿지지는 않습니다..!

현재 단계에서 직접 구분하여 구현해보는 것이 좋을까요?

sinb57 replied:
만약 Join 쿼리를 사용하지 않았다면 Repository와 Dao를 분리하는 것이 맞겠지만, 지금은 Join 쿼리도 사용중이고 미션이 간단하기 때문에 Repository와 Dao를 분리하는 것이 오히려 복잡도를 높이기 때문에 권장드리진 않아요. 다음에 한번 접해보셨으면 좋겠네요~

sinb57 commented on src/main/java/roomescape/service/reservation/MemberName.java:
```diff
@@ -0,0 +1,21 @@
+package roomescape.service.reservation;
```
`domain` 등의 패키지로 분리하지 않고 `service` 패키지 안에 위치한 이유가 궁금해요~

kjyyjk replied:
페어의 이전 패키지 구조를 그대로 따라온 것인데, 기능 구현을 하느라 미처 페어와 얘기해보지 못했었습니다.
그래서 이에 대해 제가 가지는 의도나 이유는 존재하지 않습니다..!

제대로 된 피드백을 받아보기 위해 제 방식대로 패키지 구조를 변경해 보았습니다.

계층 패키지를 나누기 이전에 가장 바깥에 도메인 별 패키지를 두었다는 점,
내부에 domain 패키지가 존재한다는 점,
AdminController와 MemberController가 /roomescape 아래로 이동되었다는 점이 변경되었습니다.

또한 `reservation`과 `theme` 와 `time`이 강하게 결합되어 있다고 판단해 하나의 패키지에서 관리하도록 했는데,
이에 대해 어떻게 생각하시는 지 궁금합니다! 
008b4e2e4196268403e994f7da5f0b43f1146176

kjyyjk replied:
하지만 이런 논리라면 결국 모든 도메인들이 하나의 패키지에서 관리되지 않을까하는 걱정도 드네요..

sinb57 replied:
지금은 각 계층 별 패키지로 분리해두었군요. 레오는 `패키지`가 무엇이라고 생각하시나요??

kjyyjk replied:
기존에 제가 생각한 `패키지`는 클래스들을 정리하는 수납함 같은 것이었습니다.
관련된 것들을 한 곳에 모아 놓고 라벨링 해놓으면 나중에 찾기 쉬우니까 패키지 구조를 잘짜야한다는 생각이었습니다.

하지만 패키지가 그 이상의 중요성을 가진다는 것을 지난 미션부터 느끼기 시작해, 이번 기회에 새로 학습해보았습니다.

패키지 구조는 컴파일된 파일들의 디렉토리 구조로 이어지며, 컴파일과 런타임 시점에 모두 사용되는 중요한 정보입니다.
만약 적절하지 않은 패키지 구조를 가져간다면 패키지 구조가 변경되었다는 이유만으로 관련 없는 클래스들도 함께
릴리즈되어야하는 문제가 생길 수 있습니다.

좋은 패키지 구조에 관한 위한 원칙들도 간략하게 알아보았고, 객체 지향 원칙들과 많이 닮아있음을 느꼈습니다.

그리고 현재는 service 패키지가 controller.dto 패키지에 의존하는데 service 패키지가 자신보다 더 변경이 쉬운 것에 의존하고 있으니 잘못된 방향의 의존성이 생긴 것 아닌가?하는 의문이 들었습니다.

학습은 [UML 실전에서는 이것만 쓴다]의 8장을 주로 참고했습니다.

sinb57 replied:
> 현재는 service 패키지가 controller.dto 패키지에 의존하는데 service 패키지가 자신보다 더 변경이 쉬운 것에 의존하고 있으니 잘못된 방향의 의존성이 생긴 것 아닌가?하는 의문이 들었습니다.

지금은 controller 패키지와 service 패키지가 양방향 의존성을 띄고 있는데요, DTO를 주로 다루는 Service 계층에서 DTO 패키지를 지니면 어떤가요?? Service 계층이 DTO 패키지를 지니면 안된다고 생각하시나요?

sinb57 commented on src/main/java/roomescape/service/reservation/Reservation.java:
```diff
@@ -1,66 +1,74 @@
-    private final ReservationTime time;
+    private final MemberName name;
+    private final ReservationDateTime dateTime;
+    private final Theme theme;
```
값객체 사용 좋네요! 👍

sinb57 commented on src/main/java/roomescape/service/ReservationTimeService.java:
```diff
@@ -1,34 +1,65 @@
+                return true;
+            }
+        }
+        return false;
```
`stream().anyMatch(..)`를 활용하여 개선해보면 어떨까요?

sinb57 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,34 +1,66 @@
-        final Reservation convertedRequest = reservationRequest.convertToReservation();
-        final Reservation reservation = reservationDao.createReservation(convertedRequest);
-        return new ReservationResponse(reservation);
+        final ReservationTime reservationTime = findReservationTime(reservationRequest.timeId());
+        final Theme theme = findTheme(reservationRequest.themeId());
+        final Reservation reservation = reservationRequest.convertToReservation(reservationTime, theme);
```
```java
        final ReservationTime reservationTime = findReservationTime(reservationRequest.timeId());
        final Theme theme = findTheme(reservationRequest.themeId());
```
위 2줄은 reservation 객체 생성을 위한 작업이네요. reservation 객체 생성에 대해 메서드 분리를 통해, 로직의 응집도를 높이고 의도를 드러내는 것은 어떨까요?

kjyyjk replied:
위 세 줄의 코드를 `makeReservation`이라는 메서드로 분리했습니다.
해당 메서드는 dao로부터 필요한 데이터를 찾아오고 `Reservation`을 만들어 반환하는 책임을 가졌습니다.
9dcfae0ea9d9e9311ce1abe64c9edbf81b3ee846

sinb57 commented on src/main/java/roomescape/service/ThemeService.java:
```diff
@@ -0,0 +1,54 @@
+    public List<ThemeResponse> findPopularThemes() {
+        final LocalDate from = LocalDate.now().minusDays(7);
+        final LocalDate to = LocalDate.now().minusDays(1);
+        final List<Theme> popularThemes = themeDao.findPopularThemes(from, to, 10);
```
매직넘버가 보이네요~ 의미를 부여해보면 어떨까요?

sinb57 commented on src/main/java/roomescape/service/reservation/ReservationTime.java:
```diff
@@ -9,16 +9,19 @@ public final class ReservationTime {
-    public ReservationTime(final Long timeId) {
-        this(timeId, null);
+    public ReservationTime(final LocalTime startAt) {
+        this(null, startAt);
```
id가 nullable 해야 하는지 궁금해요~ 0 과 같은 기본값을 사용하면 어떨까요?

kjyyjk replied:
저는 현재 도메인 객체에 위치한 id가 nullable 해야 한다고 생각합니다!

그 이유는 제가 id 필드를 완전히 db 종속적인 것으로 바라보고 있기 때문입니다.
상황에 따라 db의 식별자(id)가 0부터 시작할 가능성이 존재하기 때문에, 0과 같은 기본 값 사용은
오히려 혼란을 불러 일으킬 수 있다고 생각합니다.

kjyyjk replied:
관련해서 제 생각을 더 말씀드려볼께요!

기존에는 id 필드를 다른 name, description, startAt과 동일 선상에서 바라보려고 했으나 미션을 진행할수록
db 종속적임을 강하게 느끼고 있습니다.

이에 대해서는 크게 세 가지 근거가 존재합니다.
1. db 상의 식별자에 변경이 생기면 도메인 객체의 id 필드에도 변경이 전파된다.
2. id 필드와 다른 필드의 초기화 시점이 다르다. db에 저장되기 전까지 id 필드는 아무런 역할을 하지 못한다.
3. id 필드를 지워도 도메인 로직을 수행하는데 문제 없다.

현재 도메인 객체들은 변경되어야 할 사유가 두 개 이상이기 때문에 낮은 응집도를 가진다고 할 수 있습니다.
이에 대해서는 id 필드를 분리하고 응집도를 높일 수 있을 것 같아요.

하지만 현재 단계에서는 그렇게 얻는 이점 보다는 복잡성이 증가한다는 단점이 더 크다고 생각해서 분리하지 않았습니다.
이후 애플리케이션이 비대해지고 유지 보수성이 더 중요해진다면 분리해볼 것 같습니다.

제 생각에 대해 이프는 어떻게 생각하시는지 궁금합니다!

sinb57 replied:
레오 생각에 동의해요~ ID 식별자는 객체지향과 거리가 먼, DB 관점의 필드예요. 때문에 고민하는 부분에 크게 공감해요.

때문에 ID 식별자에 변경이 이뤄지는 경우에 대비한다면 ID 식별자 등 DB 중심적인 코드를 분리하여 관리할 수도 있겠지요. Mysql이나 H2가 아닌 MongoDB로 변경한다면, ID 식별자가 기본 String 타입이기 때문에 변경이 불가피할 수도 있어요.

하지만 웹 애플리케이션은 단순히 객체지향 만으로 작업하긴 어려워요. 그래서 Service 계층이 중요하기도 하고, 도메인도 이에 영향을 무시하진 못해서 ID 식별자 등이 강제되긴 해요. 이러한 부분은 어쩔 수 없이 받아들이고서 넘어가야 한다고 봐요. DB를 변경할 가능성이 얼마나 클지, 프레임워크를 변경할 가능성이 얼마나 클지 등에 기대하기 보다는 현재의 로직 변경이 훨씬 자주 이뤄나기 때문에 애플리케이션의 토대는 변치않을 것이라는 기대 아래 개발을 지속해야 해요. 슬프지만요.

kjyyjk replied:
레벨 1을 거쳐 현재까지 어떡하면 객체 지향적인 코드를 잘 작성할 수 있을까를 고민해왔는데,
객체 지향적이 아님을 알면서도 강제되어야 하는 부분이 있음을 체감하게 되네요..!

좋은 말씀 감사드립니다!

sinb57 commented on src/main/resources/static/js/reservation-new.js:
```diff
@@ -6,193 +6,199 @@ const timesOptions = [];
 
 document.addEventListener('DOMContentLoaded', () => {
-  document.getElementById('add-button').addEventListener('click', addInputRow);
+    document.getElementById('add-button').addEventListener('click', addInputRow);
```
formatting의 의도가 무엇인지 궁금해요~ 간혹 style formatting 때문에 PR에 files changed가 많아져서 리뷰하기 힘든 경우가 발생해요. 꼭 필요한 formatting이 아니라면 PR에 집중할 수 있도록 하지 않는 것이 좋아요. 일반적으로 작지 않은 style 변경은 별도의 PR에서 진행하곤 해요.

kjyyjk replied:
미션 진행하면서 IDEA에 의해 자동 포맷팅된 것 같습니다. 다음부터는 주의하겠습니다!

sinb57 commented on src/main/resources/static/js/reservation-new.js:
```diff
@@ -6,193 +6,199 @@ const timesOptions = [];
+    data.forEach(item => {
+        const row = tableBody.insertRow();
+
+        /*
+        TODO: [2단계] 관리자 기능 - 예약 목록 조회 API 호출 후 렌더링
+              response 명세에 맞춰 값 설정
+        */
```
반영된 요구사항이라면 제거해주세요~

sinb57 commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -86,7 +89,7 @@ void testTimeAPIs() {
                 .when().delete("/times/1")
                 .then().log().all()
-                .statusCode(200);
+                .statusCode(204);
```
숫자 코드를 사용하기 보다 HttpStatus Enum을 통해 의미를 드러내면 어떨까요?

sinb57 commented on src/test/java/roomescape/repository/ReservationDaoTest.java:
```diff
@@ -13,14 +13,17 @@
 
-    private ReservationDao reservationDao;
-    private ReservationTimeDao reservationTimeDao;
+    private H2ReservationDao reservationDao;
+    private H2ReservationTimeDao reservationTimeDao;
+    private H2ThemeDao themeDao;
```
인터페이스 타입으로 선언하는 것이 옳지 않을까 싶네요~

sinb57 commented on src/test/java/roomescape/fake/FakeThemeDao.java:
```diff
@@ -0,0 +1,61 @@
+        themes.remove(theme);
+    }
+
+    /**
+     * TODO
+     * JOIN을 구현할 방법 찾기
+     */
+    @Override
+    public List<Theme> findPopularThemes(LocalDate from, LocalDate to, int count) {
+        return null;
+    }
```
어떻게 하실지 기대되네요 👍

kjyyjk replied:
구현 방법을 몇 가지 떠올려 보았지만 이렇게 복잡하게 가져가는 것은 Fake의 영역이 아니라고 판단했습니다..ㅎ
따라서 Mockito 라이브러리의 Mock 객체를 사용해서 Service 반환 값과 메서드 호출 여부를 검증했습니다.
81440ebefe6bd2f6cc50d2f2225fa367801e0e75

sinb57 replied:
Mock 객체를 사용할 것이라면 Fake 객체가 불필요하지 않나 싶어요~ 어떻게 생각하시나요?

sinb57 reviewed (changes requested):
안녕하세요 레오~ 레벨1에 이어서 레벨2 미션으로 뵙게 되어 영광이에요ㅎ
이번 미션도 잘 부탁드려요~ 코멘트 남겨두었으니 확인해주시고 요청주세요

kjyyjk committed:
refactor: 반영된 주석 제거

kjyyjk committed:
refactor: GlobalExceptionHandler 리팩터링

- 메서드를 잘게 분리하여 응집도를 높이고 중복을 제거한다.

- 처리하지 못한 예외를 대비해 Exception을 잡아 처리하는 핸들러를 추가한다.

kjyyjk committed:
refactor: stream.anymatch를 활용한 for문 리팩터링

kjyyjk committed:
refactor: ReservationService#makeReservation 분리

- 시간과 테마를 조회해서 예약을 생성하는 것까지가 예약을 생성하는 하나의 단위라고
판단하여 메서드를 분리한다. 해당 메서드는 예약을 생성하는 책임 하나를 가지고 있어
응집도가 높다고 볼 수 있다.

kjyyjk committed:
refactor: 테스트가 Dao 인터페이스에 의존하도록 리팩터링

kjyyjk committed:
refactor: HttpStatus Enum을 활용해 Status Code 표현

kjyyjk committed:
refactor: 매직 넘버 static 변수화

kjyyjk committed:
refactor: 패키지 구조 변경

sinb57 commented on src/main/java/roomescape/controller/ThemeController.java:
```diff
@@ -0,0 +1,49 @@
+    @PostMapping("/themes")
+    public ResponseEntity<ThemeResponse> create(@RequestBody @Valid final ThemeRequest request) {
+        final ThemeResponse response = themeService.createTheme(request);
+        return ResponseEntity.created(URI.create("/thems/" + response.id())).body(response);
```
CodeRabbit이 잘 잡아줬네요!

kjyyjk replied:
앗 수정하겠습니다!

sinb57 commented on src/main/java/roomescape/repository/H2ReservationDao.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    private Reservation getReservationById(final long id) {
+        final String sql = """
```
사용하지 않는 메서드네요~

sinb57 reviewed (changes requested):
안녕하세요 레오~ 현재 미션에서 요구하는 학습 수준은 충분히 만족하셔서 이만 머지하려고 해요
코멘트 남겨놓았으니 확인해주세요! 별다른 conversation이 없다면 다음 요청 때 머지할게요~

kjyyjk committed:
refactor: 인기 테마 조회 api uri 수정

- /popular-themes -> /themes/popular-themes로 수정한다.

kjyyjk committed:
fix: 오타 수정 및 미사용 코드 제거

kjyyjk committed:
refactor: NamedParameterJdbcTemplate 적용 및 SimpleJdbcInsert 제거

- NamedParameterJdbcTemplate은 ?와 인덱스 기반이 아닌
파라미터명을 기반으로 데이터를 전달할 수 있기에 더 명확하고 안전하다.

- NamedParameterJdbcTemplate을 사용하면 inser 및 key를 반환 받는 과정도
간단해지기 때문에 현재로써는 SimpleJdbcInsert가 불필요하다고 판단했다.

kjyyjk committed:
refactor: ReservationDao의 테이블 초기화 방식 변경

- 기존에는 @BeforeEach 메서드에서 직접 drop, create 쿼리를 날리며
테이블을 초기화해주었다.

- @Sql을 사용해 실행할 script와 실행 시점을 정하도록 수정했다.
훨씬 보기 편해졌고, 스크립트를 재활용할 수 있으며 관리가 쉽다는 장점이 있다.

kjyyjk committed:
test: Dao 슬라이스 테스트

kjyyjk committed:
test: service-dao-db 통합 테스트

kjyyjk committed:
test: service#예약 가능 여부를 포함한 시간 조회 단위 테스트

kjyyjk committed:
feat: 날짜와 시간이 같아도 테마가 다르면 예약이 가능하도록 구현

kjyyjk committed:
feat: Mockito 라이브러리를 활용해 인기 테마 조회 시 Dao 모킹

kjyyjk committed:
refactor: Dao#findById의 반환 타입을 Optional로 변경

kjyyjk committed:
fix: 메서드명 오타 수정

kjyyjk committed:
refactor: FakeDao#deleteById 및 생성 후 id 검증 로직 수정

kjyyjk commented on README.md:
````diff
@@ -0,0 +1,321 @@
+HTTP/1.1 204
+```
+
+#### 예약 가능 시간 조회 API
+
+**Request**
+
+```
+GET /available-times?date=2025-03-30&themeId=1 HTTP/1.1
+```
+
+**Response**
+
+```
+HTTP/1.1 200
+Content-Type: application/json
+
+[
+    {
+        "id": 1,
+        "startAt": "10:00",
+        "alreadyBooked": false
+    }
+]
+```
````
기존에는 예약 가능 시간 조회 URI를 `/available-times`로 사용했습니다.

URI만 보고 응답값을 기대해보자면 정말 예약이 가능한 시간만을 반환할 것 같지만,
현재는 alreadyBooked true/false로 예약 가능 여부를 구분하며 예약 가능하지 않은 시간까지 전부 반환하고 있습니다.

이에 부자연스러움을 느꼈고,
바꾼다면 `/themes/{themeId}/times?date=2025-03-30`과 같이 바꿀 수 있을 것 같습니다.

하지만 이 경우 `/themes`로 시작하니 시간을 조회하는 것임에도 ThemeController에 위치하도록 변경해야하나 싶기도 합니다.

피드백 부탁드립니다..!
(본래 /times?date&themeId 로 하고 싶으나, 시간 목록 조회 api와 URI가 겹쳐서 시작된 고민입니다)

sinb57 replied:
> 하지만 이 경우 /themes로 시작하니 시간을 조회하는 것임에도 ThemeController에 위치하도록 변경해야하나 싶기도 합니다.

왜 ThemeController에 포함되면 안되는지 궁금해요~ ReservationTime 도메인은 Theme의 하위 도메인이니 문제 없지 않을까요~??

kjyyjk commented on src/main/java/roomescape/reservation/repository/H2ThemeDao.java:
```diff
@@ -0,0 +1,101 @@
+        jdbcTemplate.update(sql, parameters);
+    }
+
+    /**
+     * TODO
+     * Dao가 인기 테마라는 것을 알아야할까?
+     * Dao가 예약이 없는 테마는 인기 테마가 될 수 없다는 규칙을 알아야할까?
+     * Dao는 그냥 주어진 데이터로 db와 통신하는 책임이 아닌가?
+     */
+    @Override
+    public List<Theme> findPopularThemes(final LocalDate from, final LocalDate to, final int count) {
+        final String sql = """
+                     SELECT t.id, t.name, t.description, t.thumbnail
+                     FROM theme AS t
+                     INNER JOIN reservation AS r ON t.id = r.theme_id
+                     WHERE r.date >= :fromDate AND r.date <= :toDate
+                     GROUP BY t.id
+                     ORDER BY COUNT(*) DESC
+                     LIMIT :count
+                """;
+        final SqlParameterSource parameters = new MapSqlParameterSource()
+                .addValue("fromDate", from)
+                .addValue("toDate", to)
+                .addValue("count", count);
+        return jdbcTemplate.query(sql, parameters, themeMapper);
+    }
```
현재 sql 쿼리에는 `예약이 1건도 없는 테마는 인기 테마에 포함하지 않는다`라는 비즈니스 로직이 포함되어 있습니다.
(INNER JOIN으로 인해)
문득 여기서 dao가 비즈니스 로직까지 알아야할까라는 의문이 들었습니다.

만약 dao가 이를 몰라야한다면, Service에서 인기 순 테마를 가져온 뒤 다시 쿼리를 보내 해당 테마의 예약이 0건이라면 인기 목록에서 제외하게끔 구현해야할 것 같아요.
이로 인해 Service에 꽤 복잡하고 비효율적인 로직이 생길 것 같습니다.

현재는 그게 service와 dao가 가지는 각자의 역할과 책임이라고 생각해 후자의 복잡한 방법을 선택하는 것도 괜찮다는 생각을 하고 있습니다.

하지만 이런 논리라면 인기 테마는 예약 건수의 내림차순이라는 것 또한 비즈니스 로직이고 dao가 알고 있는 것인데, 이 부분에서는 기존과 같이 dao가 알고 있는게 적절하다고 생각됩니다.

sinb57 replied:
> dao가 비즈니스 로직까지 알아야할까라는 의문이 들었습니다

경우에 따라서는 성능 최적화를 위해 일정부분을 DB에 위임하기도 해요. 하지만 DB에 위임하는 영역에 대해 비즈니스 로직을 재빨리 파악하기 어렵기 때문에 매우 신중해야 해요.

> 하지만 이런 논리라면 인기 테마는 예약 건수의 내림차순이라는 것 또한 비즈니스 로직이고 dao가 알고 있는 것인데, 이 부분에서는 기존과 같이 dao가 알고 있는게 적절하다고 생각됩니다.

저는 정렬 쿼리가 존재해도 비즈니스 로직 상에서 한차례 더 정렬 로직을 넣곤 해요. 불필요한 연산 과정일 수 있지만, 서버 성능이 좋아진 지금은 이러한 연산 보다는 DB에 위임되어 파악이 어려운 로직이 훨씬 위험하기 때문에 Service에 의도적으로 드러내고자 작성해요. 지금 레오가 고민하는 케이스는, 쿼리를 그대로 둔 상태로 예약이 없는 경우에 대한 필터링은 Service Layer 에서 처리하면 문제 없어보여요. filtering 로직을 쿼리에 넣을 필요가 없고, 그저 10개를 가져와서 Service에서 filtering하면 되겠네욤~

kjyyjk reviewed (commented):
이번 리뷰 요청에서는 주로 테스트 코드를 보강했습니다.
`@SpringBootTest`를 활용해 service-dao를 결합한 통합 테스트도 작성했습니다.
마음 같아서는 Controller도 포함해 요청과 응답까지 테스트하고 싶었지만 다른 테스트 작성에 시간이 너무 오래 걸렸어서,
우선 리뷰 요청 보내드립니다!

sinb57 commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    private Theme findTheme(final long themeId) {
+        Optional<Theme> optionalTheme = themeDao.findById(themeId);
+        if (optionalTheme.isEmpty()) {
+            throw new IllegalArgumentException("테마가 존재하지 않습니다.");
+        }
+        return optionalTheme.get();
```
`orElseThrow`을 사용하면 어떨까요?

sinb57 reviewed (approved):
안녕하세요 레오~ 이번 단계에서 요구하는 학습 수준은 만족하셔서 이만 머지할게요! 
코멘트 확인해주시고 다음 단계 진행해주세요~ 이번 단계 고생 많으셨어요! 다음 단계도 응원할게요~
